File: classic-puzzles/puzzles-with-arrays/common_elements.py
"""
Wording:
-------
Find the common elements (as a sorted array) between two sorted arrays of integers.
"""


def common_elements(array_1: list, array_2: list) -> list:
    # Note: this solution don't use the fact that array_1 and array_2 are sorted.
    array_1_elts_in_array_2 = [elt for elt in array_1 if elt in array_2]
    common_elts_set = set(array_1_elts_in_array_2)

    return sorted(list(common_elts_set))


if __name__ == '__main__':
    array_1 = [1, 2, 4, 6, 7]
    array_2 = [2, 3, 6, 8]
    print(common_elements(array_1, array_2))

File: classic-puzzles/puzzles-with-arrays/commonElements.js
"use strict";

/*
Wording:
-------
Find the common elements (as a sorted array) between two sorted arrays of integers.
*/

/**
 * Returns common elements between two sorted arrays, as a sorted array with no duplicate.
 * @param {Array.number} array1
 * @param {Array.number} array2
 * @returns
 */
function commonElements(array1, array2) {
  let commonElts = [];
  let pointer1 = 0;
  let pointer2 = 0;

  while (pointer1 < array1.length && pointer2 < array2.length) {
    if (commonElts.includes(array1[pointer1])) {
      pointer1++;
    } else if (array1[pointer1] == array2[pointer2]) {
      commonElts.push(array1[pointer1]);
      pointer1++;
      pointer2++;
    } else if (array1[pointer1] < array2[pointer2]) {
      pointer1++;
    } else {
      pointer2++;
    }
  }
  return commonElts;
}

module.exports = commonElements;
